src/console/man/om_pseudo.1
." Text automatically generated by txt2man
.TH om_pseudo  "April 02, 2009" "" ""
.SH NAME
\fBom_pseudo \fP- generate random points (longitude and latitude pairs)
\fB
.SH SYNOPSIS
.nf
.fam C
\fBom_pseudo\fP [-] \fIv\fP \fB--version\fP | \fIn\fP \fB--num-points\fP \fINUM_POINTS\fP \fIm\fP \fB--mask\fP \fIFILE\fP [ \fB-l\fP, \fB--label\fP ] [ \fB-s\fP, \fB--seq-start\fP ] [ \fB-p\fP, \fB--proportion\fP ] [ \fB-o\fP, \fB--model\fP \fIFILE\fP \fB-t\fP, [\fB--threshold\fP \fINUM\fP] [\fB--env-unique\fP] ] [\fB--geo-unique\fP] [ \fB--log-level\fP \fILEVEL\fP ]
.fam T
.fi
.SH DESCRIPTION
\fBom_pseudo\fP is a command line tool to generate random coordinates in a specified geographic region (mask). The output will be a series of lines, one for each longitude/latitude pair, with TAB-delimited content compatible with openModeller. Coordinates will always be generated in the WGS84 datum.
By default, points can be randomly generated in any cell of the mask, except in those with "novalue". When "model" and "threshold" are specified, points will be randomly generated in any cell of the mask where the model predicts probabilities less than the threshold value. If "env-unique" is specified, all points will have unique environmental values based on the same layers that were used to create tyhe specified model. If "geo-unique" is specified, all points will have a unique pair of coordinates.
.SH OPTIONS
.TP
.B
-\fIv\fP, \fB--version\fP
Display version info.
.TP
.B
-\fIn\fP, \fB--num-points\fP
Number of points to be generated.
.TP
.B
-\fIm\fP, \fB--mask\fP
Mask file. Should be a raster file readable by openModeller. Points will never be generated in cells that contain nodata.
.TP
.B
\fB-l\fP, \fB--label\fP
Label to be associated with the points. Defaults to "label".
.TP
.B
\fB-s\fP, \fB--seq-start\fP
Id (integer) of the first point to be generated. Defaults to 1. The next points will follow the sequence.
.TP
.B
\fB-p\fP, \fB--proportion\fP
Percentage of absence points to be generated. Defaults to 100 (all points will be absence points).
.TP
.B
\fB-o\fP, \fB--model\fP
Serialized model file. When specified, points will only be generated in cells where the model predicts a probability less than the specified threshold.
.TP
.B
\fB-t\fP, \fB--threshold\fP
Probability threshold that can be used with the "model" parameter. Must be a value between 0 and 1. Default is 0.5.
.TP
.B
\fB--env-unique\fP
Can be used with the "model" parameter to avoid repeating the same environment condition for each point. This will be based on the same layers that were used to create the model.
.TP
.B
\fB--geo-unique\fP
Avoid repeating the same coordinates for each point.
.TP
.B
\fB--log-level\fP
openModeller log level: debug, warn, info or error. Defaults to "info".
.SH AUTHORS
Renato De Giovanni <renato at cria dot org dot br>


ChangeLog
Release 1.0 (2009-?-?)
------------------------
SVN revision: ?

	* Implemented new training method in the maximum entropy algorithm to produce similar results to MaxEnt.
	* Changed Artificial Neural Networks algorithm to avoid exceeding a maximum number of iterations when training by minimum error (parameter Epoch is used to specify the maximum number of iterations).
	* Added new method in the ConfusionMatrix class to use the lowest probability among the training presences as the threshold.
	* Changed the modelling protocol to allow createModel and testModel requests to specify what statistics should be calculated.
	* Removed model statistics parameters from om_model since they can be specified as part of the XML request.
	* All command line programs now return the same openModeller version number.
	* New algorithm Niche Mosaic.

Release 0.7 (2009-01-14)
------------------------
SVN revision: 4761

	* New Algorithm using Artificial Neural Networks.
	* Changed AUC calculation when no absence points are provided (instead of using 1-specificity after generating pseudo-absences, it now uses the proportion of points predicted present for a large number of random background points).
	* Included method to calculate partial area ratio in ROC class.
	* Added parameters --calc-matrix, --calc-roc, --num-background and --max-omission to om_test.
	* Changed confusion matrix serialization to store all raw values.
	* Included new map format to generate ARC/Info ASCII grid files (byte or floating point fomats).
	* Fixed bug in Maximum Entropy algorithm (environment was being re-normalized when pseudo-absences were generated internally).
	* Added support to handle categorical values in the Maximum Entropy algorithm.
	* Fixed bug in ROC curve calculation: when the corresponding algorithm required normalization and the sampler passed as a parameter was unnormalized, algorithm prediction values were being calculated based on unnormalized samples.
	* Added support to use "Spatially unique" and Environmentally unique" keywords in the request file (console interface).

Release 0.6.1 (2008-10-13)
--------------------------
SVN revision: 4580

	* Enabled integration with Condor in the modelling server.
	* Added new method in the modelling service to perform external tests.
	* Renamed om_testmodel to om_test. The new version is using getopts and accepts XML requests. A man page was also created.
	* Renamed om_create to om_model.
	* Renamed om_sampledump to om_sampler, which now accepts txt or xml input files.
	* Removed om_dump (its functionality was incorporated by the new om_sampler).
	* Added new parameter in om_points to split points into test/training with a given proportion.
	* Added new parameters in the Maximum Entropy algorithm to treat layer values in a different way (feature types).
	* Changed om_niche to receive a serialized model as a parameter.

Release 0.6.0 (2008-07-07)
--------------------------
SVN revision: 4453

	* New driver to read point data from TAPIR/DarwinCore providers.
	* New driver to read point data from GBIF.
	* New driver to read point data from serialized openModeller objects (XML).
	* New generic API for pre-analysis.
	* New console application om_algorithm replaces om_metadump.
	* New Maximum Entropy algorithm.
	* Changed input range of the GARP Best Subsets TrainingProportion parameter from 0-1 to 0-100 (percentage) and included explanation about differences between old and new GARP implementations in the algorithm description.
	* Re-enabled max threads parameter in GARP Best Subsets.
	* Removed automake files (now completely replaced by cmake).
	* Renamed compiled constant that can be used to indicate the location of algorithms from OM_LIB_PATH to OM_ALG_PATH.
	
Release 0.5.3 (2008-03-26)
--------------------------
SVN revision: 4209

	* Fixed bug when GDAL failed to read a raster row (in which case the row used to have zero values). Now the row is filled with nodata values.
	* XML request for model creation now supports additional options to filter occurrences (using spatiallyUnique or environmentallyUnique sampler functions).
	* Changed nodata value of the default raster type (ByteHFA) to 101.
	* Masks must now use nodata to indicate masked areas (zeros will not work anymore).
	* om_testmodel now generates pseudo-absence points when there are no absences to be tested.
	* Display confusion matrix cell values in om_console and om_testmodel.
	* Renamed getLayerFilename to getLayerPath and getMaskFilename to getMaskPath in EnvironmentPtr class.
	* Refactored om_pseudo, om_create and om_project to use the getopts command line library.
	* Created man pages for om_pseudo, om_create and om_project.
	* New parameter in om_pseudo to speficy the proportion of absence points to be generated.
	* New algorithm AquaMaps (for marine organisms).
	* Removed algorithms minimum distance and distance to average since EnvironmentalDistance provides the same functionality.
	* "type" property of AlgParamMetadata changed from char * to a new enumeration called AlgParamDatatype. Values can be Integer, Real and String.
	* Fixed bug in the environmental distance algorithm: probabilities were negative for points whose distance to the nearest point was beyond the maximum allowed distance. 
	* New classes to perform jackknife and chi-square in the environmental layers.
	* Updated TerraLib drivers for compatibility with TerraLib version 3.2.0.

Release 0.5.2 (2007-10-24)
--------------------------
SVN revision: 3806

	* Fixed compilation issues under Windows.
	* Included new command line program to generate pseudo occurrences.
	* Minor improvements in console tools (absences are now displayed in om_viewer and om_niche).
	* Code clean up.

Release 0.5.1 (2007-08-30)
--------------------------
SVN revision: 3661

	* Fixed MSVC compilation problems.
	* Fixed bug in deserialization of the new GARP algorithm (OM GARP).
	* Fixed crash in one-class SVM when input points contained absences.

src/console/om_pseudo.cpp
#include <openmodeller/om.hh>
#include <openmodeller/os_specific.hh>

#include "getopts/getopts.h"

#include "om_cmd_utils.hh"

#include <string>
#include <iostream>  // I/O 

#include <stdexcept>
 
using namespace std;

int main( int argc, char **argv ) {

  Options opts;
  int option;

  // command-line parameters (short name, long name, description, take args)
  opts.addOption( "" , "log-level" , "Set the log level (debug, warn, info, error)", true );
  opts.addOption( "v", "version"   , "Display the version info"                    , false );
  opts.addOption( "n", "num-points", "Number of points to be generated"            , true );
  opts.addOption( "l", "label"     , "Label for the points"                        , true );
  opts.addOption( "s", "seq-start" , "Sequence start for points id"                , true );
  opts.addOption( "m", "mask"      , "Mask file"                                   , true );
  opts.addOption( "p", "proportion", "Proportion of absence points (in %)"         , true );
  opts.addOption( "o", "model"     , "File with serialized model"                  , true );
  opts.addOption( "t", "threshold" , "Model threshold (default 0.5)"               , true );
  opts.addOption( "" , "geo-unique", "Avoid repeating same coordinates"            , false );
  opts.addOption( "" , "env-unique", "Avoid repeating same environment condition"  , false );

  std::string log_level("info");
  std::string num_points_string;
  std::string label("label");
  std::string sequence_start_string;
  std::string mask_file;
  std::string proportion_string;
  std::string model_file;
  std::string threshold_string;
  bool geo_unique = false;
  bool env_unique = false;

  if ( ! opts.parse( argc, argv ) ) {

    opts.showHelp( argv[0] ); 
    exit(0);
  }

  // Set up any related external resources
  setupExternalResources();

  OpenModeller om;

  while ( ( option = opts.cycle() ) >= 0 ) {

    switch ( option ) {

      case 0:
        log_level = opts.getArgs( option );
        break;
      case 1:
        printf( "om_pseudo %s\n", om.getVersion().c_str() );
        printf("This is free software; see the source for copying conditions. There is NO\n");
        printf("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");
        exit(0);
        break;
      case 2:
        num_points_string = opts.getArgs( option );
        break;
      case 3:
        label = opts.getArgs( option );
        break;
      case 4:
        sequence_start_string = opts.getArgs( option );
        break;
      case 5:
        mask_file = opts.getArgs( option );
        break;
      case 6:
        proportion_string = opts.getArgs( option );
        break;
      case 7:
        model_file = opts.getArgs( option );
        break;
      case 8:
        threshold_string = opts.getArgs( option );
        break;
      case 9:
        geo_unique = true;
        break;
      case 10:
        env_unique = true;
        break;
      default:
        break;
    }
  }

  // Check parameters

  if ( num_points_string.empty() ) {

    printf( "Please specify the number of points to be generated\n");
    exit(-1);
  }

  int num_points = atoi( num_points_string.c_str() );

  if ( num_points <= 0 ) {

    printf( "Please specify a valid (> 0) number of points to be generated\n");
    exit(-1);
  }

  if ( mask_file.empty() ) {

    printf( "Please specify a mask file\n");
    exit(-1);
  }
  else {

    FILE * fh = fopen( mask_file.c_str(), "r" );

    if ( fh == NULL ) {

      printf( "Could not open the specified mask file (check parameter and permissions)\n");
      exit(-1);
    }

    fclose( fh );
  }

  int sequence_start = 1; // default
  
  if ( ! sequence_start_string.empty() ) {

    sequence_start = atoi( sequence_start_string.c_str() );
  }

  if ( proportion_string.empty() ) {

    proportion_string = "100";
  }

  double proportion = atof( proportion_string.c_str() );

  if ( proportion > 100 ) {

    proportion = 100;
  }
  else if ( proportion < 0 ) {

    proportion = 0;
  }

  proportion /= 100;

  int num_absences_to_be_generated = (int)(num_points * proportion);

  double threshold = 0.5;
  
  if ( ! threshold_string.empty() ) {

    threshold = atof( threshold_string.c_str() );
  }

  if ( threshold <= 0.0 ) {

    printf( "Model threshold must be greater than zero\n");
    exit(-1);
  }

  if ( threshold >= 1.0 ) {

    printf( "Model threshold must be smaller than one\n");
    exit(-1);
  }

  // Log stuff

  Log::Level level_code = getLogLevel( log_level );

  Log::instance()->setLevel( level_code );

  // Real work

  try {

    SamplerPtr samp;

    Model model = 0;
    
    if ( model_file.empty() ) {

      std::vector<std::string> categorical_layers, continuous_layers;

      continuous_layers.push_back( mask_file ); // need to add at least one layer

      EnvironmentPtr env = createEnvironment( categorical_layers, continuous_layers, mask_file );

      if ( ! env ) {

        printf( "Could not create environment object" );
        exit(-1);
      }

      OccurrencesPtr presences( new OccurrencesImpl( label ) );
      OccurrencesPtr absences( new OccurrencesImpl( label ) );

      samp = createSampler( env, presences, absences );
    }
    else {
    
      // Load available algorithms
      AlgorithmFactory::searchDefaultDirs();

      // Load serialized model
      ConfigurationPtr config = Configuration::readXml( model_file.c_str() );

      AlgorithmPtr alg = AlgorithmFactory::newAlgorithm( config->getSubsection( "Algorithm" ) );

      model = alg->getModel();

      // note: alg deserialization doesn't include sampler stuff
      SamplerPtr alg_samp = createSampler( config->getSubsection( "Sampler" ) );

      if ( ! alg_samp ) {

        printf( "Could not find sampler data in the specified model file" );
        exit(-1);
      }

      EnvironmentPtr env = alg_samp->getEnvironment();

      env->changeMask( mask_file );

      // note: no need to change the label in presences & absences, because it 
      // will be ignored when generating the points
      OccurrencesPtr presences = alg_samp->getPresences();
      OccurrencesPtr absences = alg_samp->getAbsences();
 
      samp = createSampler( env, presences, absences );

      // Overwrite sampler, in case masks are different
      alg->setSampler( samp );

      // Normalize environment if necessary
      model->setNormalization( env );
    }

    if ( ! samp ) {

      printf( "Could not create sampler object" );
      exit(-1);
    }

    OccurrencesPtr pseudo = samp->getPseudoAbsences( num_points, model, threshold, geo_unique, env_unique );

    // Start output

    std::cerr << flush;

    // Header
    cout << "#id\t" << "label\t" << "long\t" << "lat\t" << "abundance" << endl << flush;

    string abundance("0");

    int num_generated_absences = 0;

    int i = 0;

    OccurrencesImpl::iterator it = pseudo->begin();
    OccurrencesImpl::iterator end = pseudo->end();

    // Points
    while ( it != end ) {

      // Start with absences. Switch to presence after reaching the number of absences.
      if ( num_generated_absences == num_absences_to_be_generated ) {

        abundance = "1";
      }

      std::cerr << flush;

      cout << sequence_start + i << "\t" << label.c_str() << "\t" << (*it)->x() << "\t" << (*it)->y() << "\t" << abundance.c_str() << endl << flush;

      num_generated_absences++;
      it++;
      i++;
    }
  }
  catch ( runtime_error e ) {

    printf( "om_pseudo: %s\n", e.what() );
    exit(-1);
  }

  return 0;
}
